### PR TITLE
max() uses SupportsGreaterThanT

### PR DESCRIPTION
**Pull request by srittau**
Add SupportsGreaterThan and SupportsGreaterThanT

Closes: #6336

### CODE REVIEW

**Comment by srittau** [comment]
I have no idea what the pre-commit check is trying to tell me.

**Comment by sobolevn** [comment]
Looks like it is already enabled and is looking for a config file. Which is not yet merged: https://github.com/python/typeshed/pull/6341

File: stdlib/_typeshed/__init__.pyi [hunk 1]
```diff
@@ -40,6 +40,11 @@ class SupportsLessThan(Protocol):
 
 SupportsLessThanT = TypeVar("SupportsLessThanT", bound=SupportsLessThan)  # noqa: Y001
 
+class SupportsGreaterThan(Protocol):
+    def __lt__(self, __other: Any) -> bool: ...
```
Comment (JelleZijlstra) [inline stdlib/_typeshed/__init__.pyi:44]:
```suggestion
    def __gt__(self, __other: Any) -> bool: ...
```
Important difference :)

Comment (srittau) [reply stdlib/_typeshed/__init__.pyi:44]: Whoooops.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by AlexWaygood** [comment]
I don't think the solution can be this simple, unfortunately. Consider the following class definitions:

```python
class OnlyKnowsDunderLT:
    def __init__(self, n):
        self.n = n
    def __lt__(self, other):
        return self.n < other.n
    def __repr__(self):
        return f'OnlyKnowsDunderLT(n={self.n})'

 
class OnlyKnowsDunderGT:
    def __init__(self, n):
        self.n = n
    def __gt__(self, other):
        return self.n > other.n
    def __repr__(self):
        return f'OnlyKnowsDunderGT(n={self.n})'
```

Here's how they behave with `max()` in Python 3.10:

```python
>>> max(OnlyKnowsDunderLT(5), OnlyKnowsDunderLT(6))
OnlyKnowsDunderLT(n=6)
>>> max(OnlyKnowsDunderGT(3), OnlyKnowsDunderGT(8))
OnlyKnowsDunderGT(n=8)
>>> max(OnlyKnowsDunderLT(3), OnlyKnowsDunderGT(9))
OnlyKnowsDunderGT(n=9)
```

**Comment by srittau** [comment]
@AlexWaygood That's because `__lt__` et al. are special cased by Python. Any such special cases should be special cased by type checkers as well (which they usually are).

**Comment by AlexWaygood** [comment]
@srittau, the following snippet of code [fails mypy](https://mypy-play.net/?mypy=latest&python=3.10&flags=show-error-codes%2Cstrict&gist=859dab752a4d95cc061152bc4e365308), despite running fine at runtime, indicating that `__lt__`/`__gt__` are not currently adequately special-cased by mypy. Do you think that should be reported to mypy as a bug?

```python
from typing import Protocol, Any, TypeVar

class SupportsGreaterThan(Protocol):
    def __gt__(self, other: Any) -> bool: ...
    
SupportsGreaterThanT = TypeVar('SupportsGreaterThanT', bound=SupportsGreaterThan)
    
def mymax(*iterables: SupportsGreaterThanT) -> SupportsGreaterThanT:
    return max(iterables) # type: ignore[type-var]

class HasLessThan:
    def __init__(self, n: int) -> None:
        self.n = n
    def __lt__(self, other: HasLessThan) -> bool:
        return self.n < other.n
    def __repr__(self) -> str:
        return f'HasLessThan(n={self.n})'

mymax(HasLessThan(5), HasLessThan(8))  # error: Value of type variable "SupportsGreaterThanT" of "mymax" cannot be "HasLessThan"  [type-var]
```

**Comment by srittau** [comment]
Considering that this is ultimately passed to `PyObject_RichCompareBool`, I believe this is best handled by mypy.

**Pull request merged**